controller/linodevpc_controller.go

```diff
@@ -0,0 +1,226 @@
+	vpcScope.LinodeVPC.Status.FailureReason = nil
+	vpcScope.LinodeVPC.Status.FailureMessage = util.Pointer("")
+
+	failureReason := infrav1.VPCStatusError("UnknownError")
+	//nolint:dupl // Code duplication is simplicity in this case.
+	defer func() {
+		if err != nil {
+			vpcScope.LinodeVPC.Status.FailureReason = util.Pointer(failureReason)
+			vpcScope.LinodeVPC.Status.FailureMessage = util.Pointer(err.Error())
+
+			conditions.MarkFalse(vpcScope.LinodeVPC, clusterv1.ReadyCondition, string(failureReason), clusterv1.ConditionSeverityError, err.Error())
+
+			r.Recorder.Event(vpcScope.LinodeVPC, corev1.EventTypeWarning, string(failureReason), err.Error())
+		}
+
+		if patchErr := vpcScope.PatchHelper.Patch(ctx, vpcScope.LinodeVPC); patchErr != nil && client.IgnoreNotFound(patchErr) != nil {
+			logger.Error(patchErr, "failed to patch LinodeVPC")
+
+			err = errors.Join(err, patchErr)
+		}
+	}()
+
+	// Delete
+	if !vpcScope.LinodeVPC.ObjectMeta.DeletionTimestamp.IsZero() {
+		failureReason = infrav1.DeleteVPCError
+
+		err = r.reconcileDelete(ctx, logger, vpcScope)
+
+		return
+	}
+
+	controllerutil.AddFinalizer(vpcScope.LinodeVPC, infrav1.GroupVersion.String())
+
+	// Update
+	if vpcScope.LinodeVPC.Spec.VPCID != nil {
+		failureReason = infrav1.UpdateVPCError
+
+		logger = logger.WithValues("vpcID", *vpcScope.LinodeVPC.Spec.VPCID)
+
+		err = r.reconcileUpdate(ctx, logger, vpcScope)
+
+		return
+	}
+
+	// Create
+	failureReason = infrav1.CreateVPCError
+
+	err = r.reconcileCreate(ctx, vpcScope, logger)
+
+	return
+}
+
+func (r *LinodeVPCReconciler) reconcileCreate(ctx context.Context, vpcScope *scope.VPCScope, logger logr.Logger) error {
+	if err := r.reconcileVPC(ctx, vpcScope, logger); err != nil {
+		logger.Error(err, "Failed to create VPC")
+
+		conditions.MarkFalse(vpcScope.LinodeVPC, clusterv1.ReadyCondition, string(infrav1.CreateVPCError), clusterv1.ConditionSeverityError, err.Error())
+
+		r.Recorder.Event(vpcScope.LinodeVPC, corev1.EventTypeWarning, string(infrav1.CreateVPCError), err.Error())
+
+		return err
+	}
+	vpcScope.LinodeVPC.Status.Ready = true
+
+	return nil
+}
+
+func (r *LinodeVPCReconciler) reconcileUpdate(ctx context.Context, logger logr.Logger, vpcScope *scope.VPCScope) error {
+	if err := r.reconcileVPC(ctx, vpcScope, logger); err != nil {
+		logger.Error(err, "Failed to update VPC")
+
+		conditions.MarkFalse(vpcScope.LinodeVPC, clusterv1.ReadyCondition, string(infrav1.UpdateVPCError), clusterv1.ConditionSeverityError, err.Error())
+
+		r.Recorder.Event(vpcScope.LinodeVPC, corev1.EventTypeWarning, string(infrav1.UpdateVPCError), err.Error())
+
+		return err
+	}
+	vpcScope.LinodeVPC.Status.Ready = true
+
+	return nil
+}
+
+func (r *LinodeVPCReconciler) reconcileDelete(ctx context.Context, logger logr.Logger, vpcScope *scope.VPCScope) error {
+	logger.Info("deleting VPC")
+
+	if vpcScope.LinodeVPC.Spec.VPCID != nil {
+		if err := vpcScope.LinodeClient.DeleteVPC(ctx, *vpcScope.LinodeVPC.Spec.VPCID); err != nil {
```

Comment:
think we need to do any validation the vpc is empty or anything before the delete?

Reply:
followup on this, i tested creating a vpc and deleting it with a node in it and it will error out telling you to delete the linodes... maybe this is handled and will error and reconcile gracefully?

Reply:
@luthermonson First we check vpc: https://github.com/linode/cluster-api-provider-linode/pull/62/files#diff-7fca4d96225fe1de8efb3111cc74cd194997da3af44dbde0eedf6db7cc6e257cR202 than we check subnets and if any has node(s) we re-queu the object https://github.com/linode/cluster-api-provider-linode/pull/62/files#diff-7fca4d96225fe1de8efb3111cc74cd194997da3af44dbde0eedf6db7cc6e257cR220 . After all subnets are empty we try to delete VPC. If deletion fails, we re-queue the object https://github.com/linode/cluster-api-provider-linode/pull/62/files#diff-7fca4d96225fe1de8efb3111cc74cd194997da3af44dbde0eedf6db7cc6e257cR231. I think this covers the case you have mentioned, but please feel free to fixme.

Reply:
oh i apologize, i didnt see that code come into the PR. you are correct and im good on this